writecrow/text_processing
# Add convert_to_utf8 script with semi-intelligent encoding checking

## Purpose
This is meant to resolve #2 , by providing a bit more coverage on encoding conversion. Specifically:
- Instead of just trying to open files as cp1251, it checks the character encoding, then tries to open the file with that encoding
- Allows for converting a directory of files recursively, or a single file, via arguments passed to the script
- Allows for overwriting existing files or writing to an `output` directory

## Testing steps
1. Review the [README](https://github.com/writecrow/text_processing/blob/c80a015ac4747b63c4b347e0881ae8118cddd4b3/conversion/convert_to_utf8/README.md) for thoroughness & clarity.
2. From the `convert_to_utf8` directory, run `file --mime-encoding test-data/*.txt` . You should get output verifying the variegated encoding types in the test data:

```
test-data/ascii-1.txt:        us-ascii
test-data/ascii-2.txt:        us-ascii
test-data/iso08859-1.txt:     iso-8859-1
test-data/utf-32.txt:         binary
test-data/utf-8-1.txt:        utf-8
test-data/utf16-2.txt:        utf-16be
test-data/utf16le.txt:        utf-16le
test-data/windows-1251.txt:   unknown-8bit
test-data/windows-1252-1.txt: iso-8859-1
test-data/windows-1252-2.txt: unknown-8bit
test-data/windows-1254.txt:   iso-8859-1
test-data/windows-1256.txt:   iso-8859-1
```

3. `python3 convert_to_utf8.py --directory=test-data`. Verify all test files show conversion confirmation messages:

```
test-data/.DS_Store converted from cp1252
test-data/ascii-1.txt converted from ascii
test-data/ascii-2.txt converted from ascii
test-data/iso08859-1.txt converted from latin_1
test-data/utf-32.txt converted from utf_32
test-data/utf-8-1.txt already encoded in utf-8
test-data/utf16-2.txt converted from utf-16
test-data/utf16le.txt converted from utf-16
test-data/windows-1251.txt converted from windows-1251
test-data/windows-1252-1.txt converted from latin_1
test-data/windows-1252-2.txt converted from cp1252
test-data/windows-1254.txt converted from latin_1
test-data/windows-1256.txt converted from cp1256
```

4. `file --mime-encoding output/test-data/*.txt`. Verify all files have either `ascii` or `utf-8` encoding mime types:

```
output/test-data/ascii-1.txt:        us-ascii
output/test-data/ascii-2.txt:        us-ascii
output/test-data/iso08859-1.txt:     utf-8
output/test-data/utf-32.txt:         us-ascii
output/test-data/utf-8-1.txt:        utf-8
output/test-data/utf16-2.txt:        us-ascii
output/test-data/utf16le.txt:        utf-8
output/test-data/windows-1251.txt:   utf-8
output/test-data/windows-1252-1.txt: utf-8
output/test-data/windows-1252-2.txt: utf-8
output/test-data/windows-1254.txt:   utf-8
output/test-data/windows-1256.txt:   utf-8
```
